### PR TITLE
Switch from PlotData to Entities

### LINKED ISSUE #2151: PlotData methods and structure
PlotData has a hybrid type between structs and containers.  Fix this.

### PR DESCRIPTION
**Pull request by jtlan**
<del>**EDIT**: #2147 should probably merge before this one, so I can update those API points to use `Entities` as well.</del>
**EDIT 2**: Done.

Removed `PlotData`, added the `Entity` type:
```Typescript
export type Entity = {
  datum: any;
  index: number;
  dataset: Dataset;
  position: Point;
  selection: D3.Selection;
  plot: Plot;
}
```

Renamed some methods:
* `getAllPlotData()` --> `entities()`:
```Typescript
public entities(datasets = this.datasets()): Plots.Entity[];
```
* `getClosestPlotData()` --> `entityNearest()`:
```Typescript
public entityNearest(queryPoint: Point): Plots.Entity;
```
* `plotDataAt()` --> `entitiesAt()` (`Plots.Bar` only):
```Typescript
public entitiesAt(p: Point): Entity[];
```
* `plotDataIn()` --> `entitiesIn()` (`Plots.Bar` only):
```Typescript
public entitiesIn(bounds: Bounds): Entity[];
public entitiesIn(xRange: Range, yRange: Range): Entity[];
```

Test fiddle: http://jsfiddle.net/1e7cauhd/

Close #2151.

### CODE REVIEW

File: src/components/plots/barPlot.ts [hunk 1]
```diff
@@ -169,63 +154,43 @@ export module Plots {
...
-          }
-
-          // bars minPrimaryDist away are part of the closest set
-          if (primaryDist === minPrimaryDist && secondaryDist === minSecondaryDist) {
-            closestData.push(datum);
-            closestPixelPoints.push(plotPt);
-            closestElements.push(bar);
+      var closest: Plots.Entity;
+      var entities = this.entities();
+      entities.forEach((entity) => {
```
Comment (bluong) [inline src/components/plots/barPlot.ts]: the `entities` variable can be inlined

File: src/components/plots/plot.ts [hunk 2]
```diff
@@ -13,10 +13,13 @@ module Plottable {
...
     }
 
-    export type PlotData = {
-      data: any[];
-      pixelPoints: Point[];
+    export type Entity = {
+      datum: any;
+      index: number;
+      dataset: Dataset;
+      position: Point;
```
Comment (bluong) [inline src/components/plots/plot.ts]: `pixelPosition`?

Comment (jtlan) [inline src/components/plots/plot.ts]: `Point`s are assumed to be in pixels, so adding "pixel" to the name doesn't seem helpful.

**Comment by crmorford** [comment]
About to start QE-ing, but I wanna say this is such a good starter fiddle. Thank you :)

**Comment by jtlan** [comment]
@crmorford wrote:
>About to start QE-ing, but I wanna say this is such a good starter fiddle. Thank you :)

I wasn't able to think of a way to use the new `index`, `dataset`, and `plot` parameters, but maybe you'll be able to come up with something. Maybe some sort of linked `Plot`s?

**Comment by crmorford** [comment]
http://jsfiddle.net/1e7cauhd/1/ Something's up with stacked area, try hovering over the red.

**Comment by crmorford** [comment]
it feels weird that i can can call plot.getAllPlotData().plot
is there a reason the plot is returned?

**Comment by jtlan** [comment]
@crmorford wrote:
>http://jsfiddle.net/1e7cauhd/1/ Something's up with stacked area, try hovering over the red.=

Looks like a bug from the `Drawer`s branch. @bluong is on it.

**Comment by jtlan** [comment]
@crmorford wrote:
>it feels weird that i can can call plot.getAllPlotData().plot
is there a reason the plot is returned?

Mostly a forward-looking change, in case we want `Group` to implement `entityNearest()` across any `Plot`s it contains. Another way to think of it is if you save off the `Entity` you might still want to know what `Plot` it came from.

**Comment by crmorford** [comment]
Except for the drawers bug, everything looks great!

**Comment by crmorford** [comment]
I'm confused about why entitiesIn is only for bar plot though

**Comment by jtlan** [comment]
@crmorford wrote:
>I'm confused about why entitiesIn is only for bar plot though

We don't yet have implementations for that behavior on other types of `Plot`s. #1959 is tracking that feature request.

**Pull request merged**